data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java

```diff
@@ -0,0 +1,262 @@
+        .withType(TYPE_CE_1)
+        .build();
+
+    // event sent in the response by the Callable service
+    final var responseEvent = CloudEventBuilder.v03()
+        .withId(UUID.randomUUID().toString())
+        .withDataSchema(URI.create("/api/data-schema-ce-2"))
+        .withSubject("subject-ce-2")
+        .withSource(URI.create("/api/rossi"))
+        .withData("data-ce-2".getBytes())
+        .withType(TYPE_CE_2)
+        .build();
+
+    // reconcile brokers/triggers
+    brokersManager.reconcile(Map.of(
+        new BrokerWrapper(Broker.newBuilder()
+            .setTopic(TOPIC)
+            .setId(UUID.randomUUID().toString())
+            .build()),
+        Set.of(
+            new TriggerWrapper(Trigger.newBuilder()
+                .setDestination(format("http://localhost:%d%s", SERVICE_PORT, PATH_SERVICE_1))
+                .putAttributes(ContextAttributes.TYPE.name().toLowerCase(), TYPE_CE_1)
+                .setId(UUID.randomUUID().toString())
+                .build()),
+            new TriggerWrapper(Trigger.newBuilder()
+                .setDestination(format("http://localhost:%d%s", SERVICE_PORT, PATH_SERVICE_2))
+                .putAttributes(ContextAttributes.TYPE.name().toLowerCase(), TYPE_CE_2)
+                .setId(UUID.randomUUID().toString())
+                .build())
+        )
+    ));
+
+    // start service
+    vertx.createHttpServer()
+        .exceptionHandler(context::failNow)
+        .requestHandler(request -> VertxMessageFactory.createReader(request)
+            .map(MessageReader::toEvent)
+            .onFailure(context::failNow)
+            .onSuccess(event -> {
+
+              if (request.path().equals(PATH_SERVICE_1)) { // service 1 receives requestEvent
+                context.verify(() -> {
+                  assertThat(event).isEqualTo(requestEvent);
+                  checkpoints.flag(); // 2
+                });
+
+                VertxMessageFactory.createWriter(request.response()).writeBinary(responseEvent);
+              }
+              if (request.path().equals(PATH_SERVICE_2)) { // service 2 receives responseEvent
+                context.verify(() -> {
+                  assertThat(event).isEqualTo(responseEvent);
+                  checkpoints.flag(); // 3
+                });
+              }
+            }))
+        .listen(SERVICE_PORT, "localhost");
+
+    // do the request to the Broker receiver
+    final var request = vertx.createHttpClient()
+        .post(INGRESS_PORT, "localhost", format("%s/%s", BROKER_NAMESPACE, BROKER_NAME))
+        .exceptionHandler(context::failNow)
+        .handler(response -> context.verify(() -> {
+          assertThat(response.statusCode()).isLessThan(300); // verify it's a 2xx response
+          checkpoints.flag(); // 1
+        }));
+
+    VertxMessageFactory.createWriter(request).writeBinary(requestEvent);
+  }
+
+  @AfterAll
+  public static void teardown(final VertxTestContext context) {
```

Comment:
Do you really need the vertx context here?

Reply:
If you don't have the context, the context should figure out itself to complete it